NTDownloader UI revamp

MTandi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Revamp of the software downloader program for modular computers.

Changes:
- Programs are now sorted alphabetically with incompatible ones in the end of the list.
- Installed programs are now displayed in the list.
- Added program icons.
- Moved the error messages in place of the download button.
- Only the most important error message is displayed now. Priority: compatibility, access, free space. 
- Syndicate programs are now displayed in the same list, but have a warning message from NT (There are no warning messages on syndicate OS).
- Added program categories to improve navigation. The default option "All" contains items from all categories.
- Download progress bar moved in place of the Disk usage bar. Disk usage is updated only after the download is complete, so the information was inaccurate during download. And the download bar now always visible regardless of selected category.
- The old download progress bar (next to the corresponding program) is replaced with "Downloading" indicator with a spinner.

### New UI

https://user-images.githubusercontent.com/3625094/107362787-46f0a680-6aea-11eb-9077-7ca6ad9d34a0.mp4


## Why It's Good For The Game

- The old version was hard to navigate because programs were sorted by filename.
- Error messages used vertical space and were sometimes unnecessary (I don't care about disk space or access if the program is incompatible with my device anyway)
- The old version didn't show installed programs, so you had to exit the program to see what you have installed.
- The old version grouped programs by availability. So if you found a program in a list that required access, and you inserted your ID, the list sorting was updating and you had to look for that program once again.
- Categories improve navigation when you need only software from particular category to do your job.

## Changelog
:cl:
refactor: Software Downloader program UI revamp
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MTandi committed:
initial commit

MTandi committed:
Sorting

MTandi committed:
boolean compatibility

MTandi committed:
Added categories

MTandi committed:
Tweaks according to feedback

MTandi committed:
fixed traitor access

MTandi committed:
Fixed percentage

MTandi committed:
Cleanup

MTandi committed:
Removed warning on syndi tablet

MTandi committed:
Picky linter

MTandi committed:
len to LAZYLEN

stylemistake commented:
> The old download progress bar (next to the corresponding program) is replaced with "Downloading" indicator with a spinner.

That's a downgrade :/

MTandi commented:
> That's a downgrade :/

Why? You can download only one program anyway. And I didn't remove it, I moved the progress bar to the top.

stylemistake commented:
That old progress bar looked intuitive, had a good use of space (download button was useless for the whole time it was downloadig), and would work well if someone actually added a download queue, then all of them would turn into progress bars, and fill up in queue order. Your solution is a bit of a downgrade in user experience, because it places download into free/used space area, which is counterintuitive.

stylemistake commented on tgui/packages/tgui/interfaces/NtosNetDownloader.js:
```diff
@@ -32,45 +46,49 @@ export const NtosNetDownloader = (props, context) => {
-          {downloadable_programs
-            .filter(program => program.access)
-            .map(program => (
+        <Flex>
```
Replace Flex with Stack.

Wrap Tabs with `<Section fill fitted>` element.

MTandi replied:
Thank you, I was referring to the other UI and didn't know about Stack. I'll clean things up and do something with the Downloading button.

MTandi replied:
I did not wrap tabs assuming that your other PR will be merged

stylemistake reviewed (changes requested):
I will point out more things that could've been done better, I would not merge it in this state.

MTandi committed:
Feedback processing

MTandi committed:
Changed the downloading UI logic

MTandi committed:
eslint

MTandi commented:
- Reverted the download progress bar change
- Added downloading indicator (spinner icon) to the top with a tooltip.
- Fix: Disk usage now shows real usage during download, and changed the label slightly.
- Disabled download buttons now have a tooltip

<img src="https://user-images.githubusercontent.com/3625094/107144072-b4b19c80-6949-11eb-971f-b4ba9b7e951c.png">

stylemistake commented:
Wait for #56721 to be merged, then tabs will look ok, and no actions will need to be taken.

MTandi committed:
Minor tweaks

MTandi committed:
Linting

stylemistake commented on code/modules/modular_computers/file_system/programs/ntdownloader.dm:
```diff
@@ -146,48 +154,62 @@
 				"filename" = P.filename,
 				"filedesc" = P.filedesc,
 				"fileinfo" = P.extended_desc,
+				"category" = P.category,
```
You already have category property here, there is no need to have nested lists by category, i.e. `list(category = list(...))`

stylemistake reviewed (changes requested):
I want you to do a few specific changes to your code before I can merge it.

First, sorting every second in ui_data on the server is expensive. Remove sorting on the server.

Second, simplify the code that builds the program list. You do not need to form a list for each category. Just add programs to a single, flat list, which you will then filter in TGUI based on the category property. You do not need any extra processing to the list. This will make your ui_data code very simple.

Third, this is how it's done in tgui:

```js
const programs = flow([
  // This filters the list to only contain programs with category
  // equal to selectedCategory
  filter(program => program.category === selectedCategory),
  // This sorts all programs in the lists by `name`
  sortBy(program => program.name),
])(data.programs);
```

You will need these imports:

```js
import { flow } from 'common/fp';
import { filter, sortBy } from 'common/collections';
```

Send list of categories as a simple list in ui_data:
```dm
data["categories"] = list(
  "Crew",
  "Engineering",
  ...
)
```

Do not send "All", because you can hardcode this logic into your UI, and disable the filter if "All" is selected. "All" has no purpose in dm code.

<details>
  <summary>Example of how to conditionally enable a "filter" in "flow":</summary>

```js
const programs = flow([
  selectedCategory !== 'All' && (
    filter(program => program.category === selectedCategory)
  ),
  sortBy(program => program.name),
])(data.programs);
```

</details>

Finally, rebase your PR to get new changes in master (improved Sections and Tabs), and update the screenshot. Hopefully I'm not asking too much.

MTandi committed:
Reduced BE code

MTandi committed:
Clean-up and categories fix

MTandi commented:
Updated the video in the first comment. Sorry for the confusion with rebase.

stylemistake commented on tgui/packages/tgui/interfaces/NtosNetDownloader.js:
```diff
@@ -30,11 +30,13 @@ export const NtosNetDownloader = (props, context) => {
+    && filter(program => program.verifiedsource === 1),
     // This sorts all programs in the lists by name and compatibility
-    sortBy(program => [-program.compatibility, program.filedesc]),
+    sortBy(program => [-program.compatible, program.filedesc]),
```
```suggestion
    sortBy(
      program => -program.compatible,
      program => program.filedesc),
```

MTandi committed:
Update tgui/packages/tgui/interfaces/NtosNetDownloader.js

Co-authored-by: Aleksej Komarov <stylemistake@gmail.com>

stylemistake committed:
Merge branch 'master' into ntdownloader-ui-revamp

Rohesie commented:
Does this need testing, @stylemistake ?
I see you approved it but didn't merge.
